### ISSUE TITLE
Refactor away our many many TypeScript errors

### ISSUE THREAD

**Issue by keithamoss** [ealgis/ealgis#121]
- [x] Upgrade to latest TypeScript
- [x] Research and apply best practice approaches to folder structures for Components, Actions, and Reducers (see 2DO for links)
- [x] Research the best practice approach to defining models/interfaces in React + TypeScript
- [x] Should we be passing whole objects (e.g. `mapDefinition`) around or the least amount possible (e.g. `mapDefinition.id`)?
- [x] Apply researched approaches

An example React app using material-ui:
https://github.com/TarikHuber/react-most-wanted
Has tests, components and containers, custom themeing, config, separated actions/reducers, et cetera.

# Refactoring - Stage 1
- [x] Follow Jack Hsu's approach to project structure
- [x] Apply Ducks
- [ ] ~Look at whether it's worth splitting up maps and layers into separate modules (and stores, et cetera)~ (**Decision:** Nope, not at this stage.)
- [x] Implement and use Models
- [x] Thunks and `await` (Jack Hsu)
- [x] Work out why GooglePlacesAutocomplete isn't rendering its MenuItems (may be broken since we upgraded material-ui)
- [x] Upgrade TypeScript?
- ~[ ] [@types/redux-thunk](https://github.com/gaearon/redux-thunk/issues/103) ([See also](https://www.triplet.fi/blog/type-safe-asynchronous-actions-redux-thunk-using-typescript-fsa/))~

# Refactoring - Stage 2
- [ ] [All about TypeScript 2](https://blog.mariusschulz.com/2016/09/27/typescript-2-0-non-nullable-types)
- [ ] ~Add function return types where relevant~
- [x] https://blog.mariusschulz.com/2017/06/09/typescript-2-3-the-strict-compiler-option
- [x] [Five ES6 features to make your life easier](https://www.reddit.com/r/javascript/comments/6hdsmp/five_es6_features_to_make_your_life_easier/)
- [ ] [Understanding Higher Order Components](https://www.reddit.com/r/reactjs/comments/6hso51/understanding_higher_order_components/)
- [ ] Convert our components to be 'Connected components' (i.e. Pass Ids around as much as possible, not objects - maybe [reselect](https://github.com/reactjs/reselect))

**Examples of ducks:**
- https://github.com/erikras/react-redux-universal-hot-example/blob/master/src/redux/modules/widgets.js
- https://github.com/erikras/react-redux-universal-hot-example/blob/master/src/containers/Widgets/Widgets.js
- https://github.com/erikras/react-redux-universal-hot-example/blob/master/src/components/WidgetForm/WidgetForm.js

**Research on adding async/await and Promises:**
We're targeting ES5, so there's a few hoops to jump through.

tl;dr Import `es6-promise` polyfill in `index.tsx`, configure TypeScript to inject `es2015.promise` in `tsconfig.json`

- Overview: https://blog.mariusschulz.com/2016/12/09/typescript-2-1-async-await-for-es3-es5
- Details: https://blog.mariusschulz.com/2016/11/25/typescript-2-0-built-in-type-declarations
- How to Thunk: http://ohyayanotherblog.ghost.io/redux-thunk-middleware-and-async-await/
- How to: https://medium.com/@bluepnume/learn-about-promises-before-you-start-using-async-await-eb148164a9c8

**On async/await redux-thunk:**
- https://medium.com/@zackargyle/advanced-redux-action-types-d5a71ed44e16
- https://medium.com/@jtbennett/asynchronous-actions-in-redux-8412cf92a26f
- https://basarat.gitbooks.io/typescript/docs/async-await.html

**On `Object.assign` and polyfilling:**
https://stackoverflow.com/a/40839106/7368493

**Comment by keithamoss** [comment]
**tl;dr** We'll start following Nathanael Beisiegel's approach to Feature First (Pods), perhaps combined with Ducks and dynamically adding reducers.

# Meta
- [Reddit: We're reddit's frontend engineering team. Ask us anything!](https://www.reddit.com/r/webdev/comments/5c2vie/were_reddits_frontend_engineering_team_ask_us/) (See the first post by _acemarke_)
- [React Redux Links: Project Structure](https://github.com/markerikson/react-redux-links/blob/master/project-structure.md)
- [React Redux Links: Redux Architecture and Best Practices](https://github.com/markerikson/react-redux-links/blob/master/redux-architecture.md)
- [So you’ve screwed up your Redux store — or, why Redux makes refactoring easy](https://engineering.haus.com/so-youve-screwed-up-your-redux-store-or-why-redux-makes-refactoring-easy-400e19606c71)


# Research - Project Structure Articles
- [Organising Large React Applications - Nathanael Beisiegel](http://engineering.kapost.com/2016/01/organizing-large-react-applications/)
  - [Ducks: Redux Reducer Bundles](https://github.com/erikras/ducks-modular-redux)
  - [On dynamically loading reducers](https://stackoverflow.com/questions/32968016/how-to-dynamically-load-reducers-for-code-splitting-in-a-redux-application/33044701#33044701)
- [How to scale React applications](https://www.smashingmagazine.com/2016/09/how-to-scale-react-applications/) ([Associated Twitter thread](https://twitter.com/dan_abramov/status/773912160896421889))


# Research - Project Structure Examples
- [vortigern](https://github.com/barbar/vortigern)
- [typescript-react-redux-example](https://github.com/rangle/typescript-react-redux-example)
- [react-redux-blog](https://github.com/rajaraodv/react-redux-blog)
